firmware/main.c

```diff
@@ -834,6 +834,148 @@ static void cmd_motor_mode(BaseSequentialStream *chp, int argc, char *argv[]){
 
 }
 
+/* Function for the Linear unit */
+
+static void cmd_motor_lu(BaseSequentialStream *chp, int argc, char *argv[]){
+        uint8_t i,pin2use,direction,start,gear_reduction=48,pOpt1, pOpt2; // Initialize the pins to be used //RL-D-50 has a gear reduction of 1:48
+        uint16_t num_pulses,u; // Unsigned short (0-65535)
+        char *dOpt = NULL, *pOpt = NULL, *mOpt = NULL; // *sOpt = NULL;
+        bool out1 = false, out2 = false; // Variables to store the state of the outputs
+
+        // Parsing
+        for(i = 0; i < argc; i++) {
+                if(strcmp(argv[i], "-d") == 0) {
+                        if(++i >= argc) continue;
+                        dOpt = argv[i];  // Sets the direction bit
+                }
+                else if(strcmp(argv[i], "-p") == 0) {
+                        if(++i >= argc) continue;
+                        pOpt = argv[i];  // Sets the number of pulses 
+                }
+                else if(strcmp(argv[i], "-m") == 0) {
+                        if(++i >= argc) continue;
+                        mOpt = argv[i];  // Selects the operation mode
+                }
+        /*      else if(strcmp(argv[i], "-s") == 0) {
+                        if(++i >= argc) continue;
+                        sOpt = argv[i];  // Sets the start/enable bit
+                }*/
+        }
+
+        if (!mOpt
+                || ((strcmp(mOpt, "1") == 0) && (!pOpt || !dOpt))
+                || ((strcmp(mOpt, "2") == 0) && (!pOpt || !dOpt))
+                /*|| ((strcmp(mOpt, "3") == 0) && !sOpt)*/
+                /*|| ((strcmp(mOpt, "4") == 0) && !sOpt)*/
+|| !((strcmp(mOpt, "1") == 0) || (strcmp(mOpt, "2") == 0) || (strcmp(mOpt, "3") == 0) || (strcmp(mOpt, "4") == 0))
+        )goto exit_with_usage;
+
+        /* Pin definition for the pulse and direction outputs*/
+
+        for(i = 0; i < sizeof(pinPorts)/sizeof(pinPorts[0]); i++) {
```

Comment:
All these "if-in-loop" sections look very similar, which is usually a sign for missing abstractions. That should make these more readable too.